### PR TITLE
Bug 39571: truncate long filenames

### PR DESCRIPTION
**Pull request by bvibber**
- strip more evil chars
- changes filename format to "[truncated filtered name] DD-MM-YY HH-MM-SS.jpg"
- truncates to fit filenames in 240 UTF-8 bytes, cut in half, with extra wiggle room for thumb stuff

Includes a test case with a super long monument name from Belarus.

### CODE REVIEW

File: assets/www/js/monument.js [hunk 1]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+		maxLength = maxLength / 2 - '/1024px-'.length; // halve space due to current bugs with Swift causing thumbnails to fail
+
 		var name = this.name.replace( String.fromCharCode(27), '-' );
+		name = name.replace( /[\x7f\.\[#<>\[\]\|\{\}/:]/g, '-' );
+
 		var d = date || new Date();
-		var months = [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec' ];
-		var suffix =  d.getDate() + months[ d.getMonth() ] + d.getFullYear() + ' ' + d.getHours() + 'hrs' + d.getMinutes() + 'mins' + d.getSeconds() + 'secs';
-		return name.replace( /[\x7f\.\[#<>\[\]\|\{\}]/g, '-' ) + ' (taken on ' + suffix + ').jpg';
+		function pad(str, len) {
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: whitespace

File: assets/www/js/monument.js [hunk 2]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+			if ( typeof str != 'string' ) {
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: !==

File: assets/www/js/monument.js [hunk 3]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+			'.jpg';
+
+		var allowedLength = maxLength - suffix.length;
+
+		// Count UTF-8 bytes to see where we need to crop long names.
+		var bytes = 0, chars = 0;
+		for ( var i = 0; i < name.length; i++ ) {
+			// JavaScript strings are UTF-16.
+			var codeUnit = name.charCodeAt( i );
+			var len;
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: len already declared

File: assets/www/js/monument.js [hunk 4]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+			if ( codeUnit < 0x80 ) {
+				len = 1;
+			} else if ( codeUnit < 0x800 ) {
+				len = 2;
+			} else if ( codeUnit >= 0xd800 && codePoint < 0xe000 ) {
+				// http://en.wikipedia.org/wiki/UTF-16#Description
+				// Code point is one half of a surrogate pair.
+				// This and its partner combine to form a single 4 byte character in UTF-8.
+				len = 4;
+				i++;
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: this feels wrong

File: assets/www/js/monument.js [hunk 5]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+			if ( bytes + len < allowedLength ) {
+				bytes += len;
+				chars++;
+				if ( len == 4 ) {
+					// Skip over second half of surrogate pair as a unit.
+					chars++;
+					i++;
+				}
+			} else {
+				break;
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: avoid breaks in loops.

File: assets/www/js/monument.js [hunk 6]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+			}
+		}
+
+		var base = name.substr( 0, chars );
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: i'd just combine this line with second

File: assets/www/js/monument.js [hunk 7]
```diff
@@ -33,13 +33,76 @@ define( [ 'jquery', 'utils' ], function() {
...
+		return base + suffix;
 	};
-	
+
```
Comment (jdlrobson) [inline assets/www/js/monument.js]: unrelated whitespace

**Comment by jdlrobson** [comment]
Looks good but a few bits of refactoring (and please do these as fixups during a rebase) before merging please.

File: assets/www/js/monument.js (hunk 4 above)
Comment (bvibber) [inline assets/www/js/monument.js]: woops yeah that i++ shouldn't be there, it's covered in the "skip over second half" below.

File: assets/www/js/monument.js (hunk 5 above)
Comment (bvibber) [inline assets/www/js/monument.js]: What would you recommend? I could break out the utf-8 trimming to a separate function and use 'return' instead of 'break', but that's pretty much exactly the same thing. :)

Comment (bvibber) [inline assets/www/js/monument.js]: I've gone ahead and split it out to a separate function in utils.js, also fixed soem bugs in it. :)
